tests/apollo/test_skvbc_backup_restore.py

```diff
@@ -0,0 +1,134 @@
+    persistence enabled (-p).
+
+    Note each arguments is an element in a list.
+    """
+    statusTimerMilli = "500"
+    viewChangeTimeoutMilli = "10000"
+    path = os.path.join(builddir, "tests", "simpleKVBC", "TesterReplica", "skvbc_replica")
+    return [path,
+            "-k", KEY_FILE_PREFIX,
+            "-i", str(replica_id),
+            "-s", statusTimerMilli,
+            "-v", viewChangeTimeoutMilli,
+            "-p",
+            "-t", os.environ.get('STORAGE_TYPE')]
+
+
+class SkvbcBackupRestoreTest(unittest.TestCase):
+
+    __test__ = False  # so that PyTest ignores this test scenario
+
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    async def test_checkpoint_propagation_after_restarting_replicas(self, bft_network):
+        """
+        Here we trigger a checkpoint, restart all replicas in a random order with 10s delay in-between,
+        both while stopping and starting. We verify checkpoint persisted upon restart and then trigger
+        another checkpoint. We make sure checkpoint is propagated to all the replicas.
+
+        1) Given a BFT network, we make sure all nodes are up
+        2) Send sufficient number of client requests to trigger checkpoint protocol
+        3) Stop all replicas in a random order (with 10s delay in between)
+        4) Start all replicas in a random order (with 10s delay in between)
+        5) Send sufficient number of client requests to trigger another checkpoint
+        6) Make sure checkpoint propagates to all the replicas
+        7) Make sure the initial view is stable
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        n = bft_network.config.n
+
+        self.assertEqual(len(bft_network.procs), n, "Make sure all replicas are up initially.")
+
+        current_primary = await bft_network.get_current_primary()
+
+        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=current_primary)
+
+        await skvbc.fill_and_wait_for_checkpoint(
+            initial_nodes=bft_network.all_replicas(),
+            checkpoint_num=1,
+            verify_checkpoint_persistency=False
+        )
+
+        # verify checkpoint propagation to all the nodes
+        for replica in bft_network.all_replicas():
+            first_checkpoint_after = await bft_network.wait_for_checkpoint(
+                replica_id=replica,
+                expected_checkpoint_num=checkpoint_before + 1)
+            self.assertEqual(first_checkpoint_after, checkpoint_before + 1)
+
+        # stop n replicas in a random order with a delay of 10s in between
+        stopped_replicas = await self._stop_random_replicas_with_delay(bft_network, n)
+
+        # start stopped replicas in a random order with a delay of 10s in between
+        await self._start_random_replicas_with_delay(bft_network, stopped_replicas)
+
+        # verify checkpoint persisted in all the nodes after restart
```

Comment:
You can call `bft_network.wait_for_replicas_to_checkpoint(bft_network.all_replicas(), checkpoint_before + 1)` to run the checks in parallel rather than having to write the loop manually.

Reply:
Done for the same reason as above.